bioconda_utils/lint/__init__.py

```diff
@@ -0,0 +1,446 @@
+   :toctree:
+
+   checks
+
+"""
+
+import abc
+import os
+import re
+import itertools
+import logging
+from collections import defaultdict
+from enum import IntEnum
+from typing import Any, Dict, List, NamedTuple, Tuple
+
+import pandas as pd
+import ruamel_yaml as yaml
+import networkx as nx
+
+from .. import utils
+from ..recipe import Recipe, RecipeError
+
+logger = logging.getLogger(__name__)
+
+
+class Severity(IntEnum):
+    """Severities for lint checks"""
+    #: Checks of this severity are purely informational
+    INFO = 10
+
+    #: Checks of this severity indicate possible problems
+    WARNING = 20
+
+    #: Checks of this severity must be fixed and will fail a recipe.
+    ERROR = 30
+
+INFO = Severity.INFO
+WARNING = Severity.WARNING
+ERROR = Severity.ERROR
+
+
+class LintMessage(NamedTuple):
+    """Message issued by LintChecks"""
+
+    #: The recipe this message refers to
+    recipe: Recipe
+
+    #: The check issuing the message
+    check: 'LintCheck'
+
+    #: The severity of the message
+    severity: Severity = ERROR
+
+    #: Message title to be presented to user
+    title: str = ""
+
+    #: Message body to be presented to user
+    body: str = ""
+
+    #: Line at which problem begins
+    start_line: int = 0
+
+    #: Line at which problem ends
+    end_line: int = 0
+
+    #: Name of file in which error was found
+    fname: str = "meta.yaml"
+
+    def get_level(self):
+        """Return level string as required by github"""
+        if self.severity < WARNING:
+            return "notice"
+        if self.severity < ERROR:
+            return "warning"
+        return "failure"
+
+
+class LintCheckMeta(abc.ABCMeta):
+    """Meta class for lint checks
+
+    Handles registry
+    """
+
+    registry: List["LintCheck"] = []
+
+    def __new__(cls, name: str, bases: Tuple[type, ...],
+                namespace: Dict[str, Any], **kwargs) -> type:
+        """Creates LintCheck classes"""
+        typ = super().__new__(cls, name, bases, namespace, **kwargs)
+        if name != "LintCheck":  # don't register base class
+            cls.registry.append(typ)
+        return typ
+
+    def __str__(cls):
+        return cls.__name__
+
+
+def get_checks():
+    """Returns the registered lint checks"""
+    return LintCheckMeta.registry
+
+
+class LintCheck(metaclass=LintCheckMeta):
+    """Base class for lint checks"""
+
+    #: Severity of this check. Only ERROR causes a lint failure.
+    severity: Severity = ERROR
+
+    #: Checks that must have passed for this check to be executed.
+    requires: List['LintCheck'] = []
+
+    def __init__(self, _linter: 'Linter') -> None:
+        self.messages: List[LintMessage] = []
+        self.recipe: Recipe = None
+
+    def __str__(self):
+        return self.__class__.__name__
+
+    def run(self, recipe: Recipe) -> List[LintMessage]:
+        """Run the check on a recipe. Called by Linter"""
+        self.messages: List[LintMessage] = []
+        self.recipe: Recipe = recipe
+
```

Comment:
It's not clear what happens to the whole linting system when a check raises an exception. The current checks are pretty simple (hopefully that means robust), but it would be good to gracefully handle exceptions. Maybe wrap each `self.check_*()` with try/except and if exception then a message gets added saying the process of checking failed?

Reply:
Also done.